### PR TITLE
feat: Add `ActixMessageWrapper` to rate limit `PeerActor`

### PR DESCRIPTION
**Pull request by pmnoxx**
We want to track of sizes of messages sent from `PeerActor` to `PeerManagerActor`.
We will stop reading from given `PeerActor` once we reach the limit.

To achieve this a new type `ActixMessageWrapper` was introduced. It wraps around messages sent to peer, and contains metadata needed for stat counting.

This is a draft:
- [x] Add a wrapper for sending messages from`PeerActor` to `PeerManagerActor`
- [x] Modify `PeerActor` to use 
- [x] Add a wrapper for reply's to the above messages (wating for `DeepSizeOf`)
- [ ] Wait for `DeepSizeOf` (another PR)
- [ ] Add a wrapper for messages to Peer - for example for routing (another PR)
- [ ] Add a wrapper for messages to EdgeVerifierActor - for example for routing (another PR)
- [ ] Add a wrapper for messages to RoutingTableActor - for example for routing (another PR)
- [ ] Add a wrapper for messages to ClientActor - for example for routing (another PR)
- [ ] Extensions to `performance_stats` to track per thread limits (another PR)
- [ ] Extension to track bandwidth (another PR)

This PR is based on, not yet merged, https://github.com/near/nearcore/pull/4927

Resolves https://github.com/near/nearcore/issues/5155

Blocked by https://github.com/Aeledfyr/deepsize/pull/18

### CODE REVIEW

File: chain/network/src/common/message_wrapper.rs
Comment (matklad) [inline chain/network/src/common/message_wrapper.rs:57]: More idiomatic name would be `into_inner`

**Review by matklad — COMMENTED** [review]
Alternative design which might or might not work: have only a single wrapper type and add delegating impls:

```rust
pub struct Throttled<T> {
    inner: T,
    throttle_controller: ThrottleController,
    msg_len: usize,
}

impl<M: actix::Message> actix::Message for Throttled<M> {
    type Result = Throttled<M::Result>;
}

impl<T> Throttled<T> {
    pub fn new(inner: T, throttle_controller: ThrottleController, msg_len: usize) -> Self {
        throttle_controller.add_msg(msg_len);
        Self { inner, throttle_controller, msg_len }
    }

    pub fn into_inner(mut self) -> T {
        self.throttle_controller.remove_msg(self.msg_len);
        self.inner
    }
}

impl<T: actix::Message> Throttled<T> {
    pub fn into_response(self, resp: T::Result) -> Throttled<T::Result> {
        Throttled {
            inner: resp,
            throttle_controller: self.throttle_controller,
            msg_len: self.msg_len,
        }
    }
}

impl<T> ops::Deref for Throttled<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
```

File: chain/network/src/common/message_wrapper.rs
Comment (matklad) [inline chain/network/src/common/message_wrapper.rs:24]: `rate_limiter: ThrottleController`

Let's use consistent terminology: I think this' relatively important that near_xxx crate, XXXFrameReader, XXController and `xxx: XXXContrloller` all use the same known for `xxx`.

**Comment by matklad** [comment]
:thinking: actually, I am not sure that we need wrappers here: you can make them work using the wonders of Rust type-system but they are awkward to use. Perhaps a more ergonomic variant would be to define 

```rust
// NB: deliberately not `Clone`
struct ThrottleToken {
    throttle_controller: ThrottleController,
    msg_len: usize,
}

impl ThrottleToken {
    fn new(mut throttle_controller: ThrottleController, msg_len: usize) -> Self {
        throttle_controller.add_msg(msg_len);
        Self { throttle_controller, msg_len }
    }
}

impl Drop for ThrottleToken {
    fn drop(&mut self) {
        self.throttle_controller.remove_msg(self.msg_len)
    }
}
```

and then just add it as a field to every request and response? The absence of `Clone` guarantees that the token would be threaded correctly from requests to responses.

File: chain/network/src/common/message_wrapper.rs
Comment (pmnoxx) [reply chain/network/src/common/message_wrapper.rs]: Good idea, I renamed package to `near-throttled`

**Comment by pmnoxx** [comment]
> thinking actually, I am not sure that we need wrappers here: you can make them work using the wonders of Rust type-system but they are awkward to use. Perhaps a more ergonomic variant would be to define
> 
> ```rust
> // NB: deliberately not `Clone`
> struct ThrottleToken {
>     throttle_controller: ThrottleController,
>     msg_len: usize,
> }
> 
> impl ThrottleToken {
>     fn new(mut throttle_controller: ThrottleController, msg_len: usize) -> Self {
>         throttle_controller.add_msg(msg_len);
>         Self { throttle_controller, msg_len }
>     }
> }
> 
> impl Drop for ThrottleToken {
>     fn drop(&mut self) {
>         self.throttle_controller.remove_msg(self.msg_len)
>     }
> }
> ```
> 
> and then just add it as a field to every request and response? The absence of `Clone` guarantees that the token would be threaded correctly from requests to responses.

Right now we are in the middle of doing the transition. So, we have two handlers, one that takes the message by as it is, and another one with the wrapper that contains `ThrottleToken`. We probably won't be able to avoid to for now.

**Comment by pmnoxx** [comment]
@mm-near @matklad 
In order for this change not to be blocked by `DeepSizeOf`. I simplified the change, `ActixMessageWrapper` will not care about messages sizes anymore just the count.

That should still serve the intended function. 

We can put `DeepSizeOf` under a feature flag, not compiled by default, and enabled only when needed for now.

File: chain/network/src/common/message_wrapper.rs
Comment (matklad) [inline chain/network/src/common/message_wrapper.rs:15]: I'd say that one type parameter should be enough here: 

```diff
diff --git a/chain/network/src/common/message_wrapper.rs b/chain/network/src/common/message_wrapper.rs
index ad66e948c..d96a909a2 100644
--- a/chain/network/src/common/message_wrapper.rs
+++ b/chain/network/src/common/message_wrapper.rs
@@ -12,15 +12,14 @@ use std::marker::PhantomData;
 /// - Add needed decorators. Probably `Debug`, `Message` from Actix, etc.
 /// - Add two rate limiters (local per peer, global one)
 /// - Any other metadata we need debugging, etc.
-pub struct ActixMessageWrapper<T, Q> {
+pub struct ActixMessageWrapper<T> {
     msg: T,
     throttle_token: ThrottleToken,
-    ph: PhantomData<Q>,
 }
 
-impl<T, Q> ActixMessageWrapper<T, Q> {
+impl<T> ActixMessageWrapper<T> {
     pub fn new_without_size(msg: T, throttle_controller: ThrottleController) -> Self {
-        Self { msg, throttle_token: ThrottleToken::new(throttle_controller, 0), ph: PhantomData }
+        Self { msg, throttle_token: ThrottleToken::new(throttle_controller, 0) }
     }
 
     #[allow(unused)]
@@ -34,11 +33,8 @@ impl<T, Q> ActixMessageWrapper<T, Q> {
     }
 }
 
-impl<T, Q> Message for ActixMessageWrapper<T, Q>
-where
-    Q: 'static,
-{
-    type Result = ActixMessageResponse<Q>;
+impl<T: Message> Message for ActixMessageWrapper<T> {
+    type Result = ActixMessageResponse<T::Result>;
 }
 
 #[derive(MessageResponse)]
diff --git a/chain/network/src/peer_manager/peer_manager_actor.rs b/chain/network/src/peer_manager/peer_manager_actor.rs
index 4cd7f2ec9..1b58ef69f 100644
--- a/chain/network/src/peer_manager/peer_manager_actor.rs
+++ b/chain/network/src/peer_manager/peer_manager_actor.rs
@@ -2212,14 +2212,12 @@ impl PeerManagerActor {
     }
 }
 
-impl Handler<ActixMessageWrapper<PeerManagerMessageRequest, PeerManagerMessageResponse>>
-    for PeerManagerActor
-{
+impl Handler<ActixMessageWrapper<PeerManagerMessageRequest>> for PeerManagerActor {
     type Result = ActixMessageResponse<PeerManagerMessageResponse>;
 
     fn handle(
         &mut self,
-        msg: ActixMessageWrapper<PeerManagerMessageRequest, PeerManagerMessageResponse>,
+        msg: ActixMessageWrapper<PeerManagerMessageRequest>,
         ctx: &mut Self::Context,
     ) -> Self::Result {
         // Unpack throttle controller
```

File: chain/network/src/common/message_wrapper.rs
Comment (matklad) [inline chain/network/src/common/message_wrapper.rs:27]:
```suggestion
    pub fn into_inner(mut self) -> T {
```

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (matklad) [inline chain/network/src/peer_manager/peer_manager_actor.rs:2231]: We can write this logic once in a generic way: 

```diff
diff --git a/chain/network/src/common/message_wrapper.rs b/chain/network/src/common/message_wrapper.rs
index ad66e948c..b2c028da1 100644
--- a/chain/network/src/common/message_wrapper.rs
+++ b/chain/network/src/common/message_wrapper.rs
@@ -1,5 +1,5 @@
 use actix::dev::MessageResponse;
-use actix::Message;
+use actix::{Actor, Handler, Message};
 use near_rate_limiter::{ThrottleController, ThrottleToken};
 use std::marker::PhantomData;
 
@@ -12,15 +12,14 @@ use std::marker::PhantomData;
 /// - Add needed decorators. Probably `Debug`, `Message` from Actix, etc.
 /// - Add two rate limiters (local per peer, global one)
 /// - Any other metadata we need debugging, etc.
-pub struct ActixMessageWrapper<T, Q> {
+pub struct ActixMessageWrapper<T> {
     msg: T,
     throttle_token: ThrottleToken,
-    ph: PhantomData<Q>,
 }
 
-impl<T, Q> ActixMessageWrapper<T, Q> {
+impl<T> ActixMessageWrapper<T> {
     pub fn new_without_size(msg: T, throttle_controller: ThrottleController) -> Self {
-        Self { msg, throttle_token: ThrottleToken::new(throttle_controller, 0), ph: PhantomData }
+        Self { msg, throttle_token: ThrottleToken::new(throttle_controller, 0) }
     }
 
     #[allow(unused)]
@@ -34,11 +33,20 @@ impl<T, Q> ActixMessageWrapper<T, Q> {
     }
 }
 
-impl<T, Q> Message for ActixMessageWrapper<T, Q>
-where
-    Q: 'static,
-{
-    type Result = ActixMessageResponse<Q>;
+impl<T: Message> Message for ActixMessageWrapper<T> {
+    type Result = ActixMessageResponse<T::Result>;
+}
+
+impl<T: Message> ActixMessageWrapper<T> {
+    pub fn handle<A: Actor + Handler<T, Result = T::Result>>(
+        self,
+        actor: &mut A,
+        ctx: &mut A::Context,
+    ) -> ActixMessageResponse<T::Result> {
+        let (msg, throttle_token) = self.take();
+        let result = actor.handle(msg, ctx);
+        ActixMessageResponse::new(result, ThrottleToken::new(throttle_token.into_inner(), 0))
+    }
 }
 
 #[derive(MessageResponse)]
diff --git a/chain/network/src/peer_manager/peer_manager_actor.rs b/chain/network/src/peer_manager/peer_manager_actor.rs
index 4cd7f2ec9..0808a6246 100644
--- a/chain/network/src/peer_manager/peer_manager_actor.rs
+++ b/chain/network/src/peer_manager/peer_manager_actor.rs
@@ -2212,23 +2212,15 @@ impl PeerManagerActor {
     }
 }
 
-impl Handler<ActixMessageWrapper<PeerManagerMessageRequest, PeerManagerMessageResponse>>
-    for PeerManagerActor
-{
+impl Handler<ActixMessageWrapper<PeerManagerMessageRequest>> for PeerManagerActor {
     type Result = ActixMessageResponse<PeerManagerMessageResponse>;
 
     fn handle(
         &mut self,
-        msg: ActixMessageWrapper<PeerManagerMessageRequest, PeerManagerMessageResponse>,
+        msg: ActixMessageWrapper<PeerManagerMessageRequest>,
         ctx: &mut Self::Context,
     ) -> Self::Result {
-        // Unpack throttle controller
-        let (msg, throttle_token) = msg.take();
-
-        let result = self.handle(msg, ctx);
-
-        // TODO(#5155) Add support for DeepSizeOf to result
-        ActixMessageResponse::new(result, ThrottleToken::new(throttle_token.into_inner(), 0))
+        msg.handle(self, ctx)
     }
 }
 
```

Not sure if we should hough -- in general, I personally don't like such type-heavy code.

Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:2231]: @matklad It's good suggestion. The answer is as always it depends. If it's only one use, then I wouldn't bother.
In the next PR, I'll include need more wrappers. We can decide what do it at that point.

Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:2229]: Yes, this works. However, I'll in the next PR I'll have to pass `throttle_controller` from ThrottleToken inside `actor.handle()`.  This generic will not work, I'll have to use something different.

```
-> ActixMessageResponse<T::Result> {
+        let (msg, throttle_token) = self.take();
+        let result = actor.handle(msg, ctx);
+        ActixMessageResponse::new(result, ThrottleToken::new(throttle_token.into_inner(), 0))
+    }
```

**Comment by pmnoxx** [comment]
@mm-near Can you take a look?

**Review by matklad — APPROVED** [review]

**Comment by pmnoxx** [comment]
@mm-near Can you take a look?

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:107]: do you want to keep it at 0 ?

**Review by mm-near — APPROVED** [review]

**Comment by pmnoxx** [comment]
I'll merge after https://github.com/near/nearcore/pull/5191

**Comment by frol** [comment]
I don't have capacity to review this PR now. If you believe I should still do, let me know, otherwise feel free to make your own call.

**Pull request merged**
